libs/coin-evm/src/api/node/ledger.common.ts

```diff
@@ -0,0 +1,352 @@
+  const { node } = currency.ethereumLikeInfo || {};
+  if (!isLedgerNodeConfig(node)) {
+    throw new LedgerNodeUsedIncorrectly();
+  }
+
+  const { balance } = await fetchWithRetries<{ address: string; balance: string }>({
+    method: "GET",
+    url: `${getEnv("EXPLORER")}/blockchain/v4/${node.explorerId}/address/${address}/balance`,
+  });
+
+  return new BigNumber(balance);
+};
+
+/**
+ * Get the balance of multiples tokens for addresses
+ */
+export const getBatchTokenBalances = async (
+  input: { address: string; contract: string }[],
+  params: { currency: CryptoCurrency },
+): Promise<BigNumber[]> => {
+  const { currency } = params;
+  const { node } = currency.ethereumLikeInfo || {};
+  if (!isLedgerNodeConfig(node)) {
+    throw new LedgerNodeUsedIncorrectly();
+  }
+
+  const balances = await fetchWithRetries<
+    Array<{
+      address: string;
+      contract: string;
+      balance: string;
+    }>
+  >({
+    method: "POST",
+    url: `${getEnv("EXPLORER")}/blockchain/v4/${node.explorerId}/erc20/balances`,
+    data: input,
+  });
+
+  return balances.map(({ balance }) => new BigNumber(balance));
+};
+
+/**
+ * Get the balance of an address for an ERC20 token
+ */
+export const getTokenBalance: RpcApi["getTokenBalance"] = async (
+  currency,
+  address,
+  contractAddress,
+) => {
+  const { node } = currency.ethereumLikeInfo || {};
+  if (!isLedgerNodeConfig(node)) {
+    throw new LedgerNodeUsedIncorrectly();
+  }
+
+  if (!tokenBalancesBatchersMap.has(currency)) {
+    const batcher = makeBatcher(getBatchTokenBalances, { currency });
+    tokenBalancesBatchersMap.set(currency, batcher);
+  }
+  const requestBatcher = tokenBalancesBatchersMap.get(currency)!;
+
+  return requestBatcher({ address, contract: contractAddress });
+};
+
+/**
+ * Get account nonce
+ */
+export const getTransactionCount: RpcApi["getTransactionCount"] = async (currency, address) => {
+  const { node } = currency.ethereumLikeInfo || {};
+  if (!isLedgerNodeConfig(node)) {
+    throw new LedgerNodeUsedIncorrectly();
+  }
+
+  const { nonce } = await fetchWithRetries<{
+    address: string;
+    nonce: number;
+  }>({
+    method: "GET",
+    url: `${getEnv("EXPLORER")}/blockchain/v4/${node.explorerId}/address/${address}/nonce`,
+  });
+
+  return nonce;
+};
+
+/**
+ * Get an estimated gas limit for a transaction
+ */
+export const getGasEstimation: RpcApi["getGasEstimation"] = async (account, transaction) => {
+  const { currency } = account;
+  const { node } = currency.ethereumLikeInfo || {};
+  if (!isLedgerNodeConfig(node)) {
+    throw new LedgerNodeUsedIncorrectly();
+  }
+
+  const { recipient: to, amount: value, data } = transaction;
+
+  try {
+    const { estimated_gas_limit: gasEtimation } = await fetchWithRetries<{
+      to: string;
+      estimated_gas_limit: string;
+    }>({
+      method: "POST",
+      url: `${getEnv("EXPLORER")}/blockchain/v4/${node.explorerId}/tx/estimate-gas-limit`,
```

Comment:
We gotta be careful and reproduce the bug that happened with ethereum family (the last v3 endpoint we are using)

Reply:
Good point, but we need to recall what the pb was with the v4 endpoint at the time. It just didn't exist right? But now that it does I don't know if there was any specific bug besides?

Reply:
There was no bug, I'm not sure to understand what you mean @hedi-edelbloute. Here we're simply not using the v3 endpoint as it wasn't really needed in the first place. 🤔

Reply:
I digged in Jira and there was actually two problems : one on compound, and one when swapping ETH for XRP for example

Reply:
Like I said, all of those cases are covered by the fact that if a transaction simulation fails, it's catched at the prepareTransaction level and `gasLimit` is set at 0 which is detected as an error in the `getTransactionStatus` level.